src/communication/websocket/session.cpp

```diff
@@ -0,0 +1,101 @@
+// Copyright 2021 Memgraph Ltd.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt; by using this file, you agree to be bound by the terms of the Business Source
+// License, and you may not use this file except in compliance with the Business Source License.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+#include "communication/websocket/session.hpp"
+
+#include "utils/logging.hpp"
+
+namespace communication::websocket {
+namespace {
+void LogError(boost::beast::error_code ec, const std::string_view what) {
+  spdlog::warn("Websocket session failed on {}: {}", what, ec.message());
+}
+}  // namespace
+
+void Session::Run() { OnRun(); }
+
+void Session::Write(const std::string_view message) {
+  if (!connected_.load(std::memory_order_relaxed)) {
+    return;
+  }
+  const auto message_string = std::make_shared<std::string>(message);
+  boost::asio::dispatch(ws_.get_executor(), [message_string, shared_this = shared_from_this()] {
+    shared_this->messages_.push_back(message_string);
+
+    if (shared_this->messages_.size() > 1) {
+      return;
+    }
+    shared_this->DoWrite();
+  });
+}
+
+bool Session::Connected() { return connected_.load(std::memory_order_relaxed); }
+
+void Session::DoWrite() {
+  const auto next_message = messages_.front();
+  ws_.async_write(boost::asio::buffer(*next_message), [message_string = next_message, shared_this = shared_from_this()](
```

Comment:
There is way to avoid an allocation here. I would move the element from the front to the lambda (as a sink) and in the case of `if(ec)` on `OnWrite` I would move it back at the front.

Reply:
What allocation exactly? It's a `shared_ptr` so nothing much should happen?

Reply:
I am talking about the string copy, `message_string = next_message`

Reply:
`next_message` is a `shared_ptr` so no string copy going on here.

Reply:
Actually copying a shared_ptr requires locking (to modify the ref counter), so move does make sense here in my opinion.